Add more update rules

abcdefg30 opened the pull request:
Another step towards completing #14600.

pchote committed:
Add two more update rules

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/AddNukeLaunchAnimation.cs:
```diff
@@ -0,0 +1,42 @@
+{
+	public class AddNukeLaunchAnimation : UpdateRule
+	{
+		public override string Name { get { return "Add 'WithNukeLaunchAnimation' and remove 'ActivationSequence'"; } }
```
These should mention `NukePower.ActivationSequence` so users know what trait you are talking about.

pchote commented on OpenRA.Mods.Common/UpdateRules/UpdatePath.cs:
```diff
@@ -34,6 +34,19 @@ public class UpdatePath
 			Justification = "Extracting update lists to temporary variables obfuscates the definitions.")]
 		static readonly UpdatePath[] Paths =
 		{
+			new UpdatePath("release-20171014", "release-20180218", new UpdateRule[]
```
Can we please comment out this block or maybe rename it to `incomplete-release-20171014` to make it clear that this is not ready yet?

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/AddNukeLaunchAnimation.cs:
```diff
@@ -0,0 +1,42 @@
+		{
+			foreach (var nuke in actorNode.ChildrenMatching("NukePower"))
+			{
+				var sequence = nuke.LastChildMatching("ActivationSequence").Value.Value;
```
`.NodeValue<string>();`

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/CapturableChanges.cs:
```diff
@@ -0,0 +1,86 @@
+			var allowAllies = node.LastChildMatching("AllowAllies");
+			if (allowAllies != null)
+			{
+				if (FieldLoader.GetValue<bool>("AllowAllies", allowAllies.Value.Value))
```
`if (allowAllies.NodeValue<bool>())` and so on.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/CapturableChanges.cs:
```diff
@@ -0,0 +1,86 @@
+			}
+
+			if (stance != (Stance.Neutral | Stance.Enemy))
+				actorNode.AddNode("ValidStances", stance.ToString());
```
`actorNode.AddNode("ValidStances", stance);` (the `ToString` isn't needed).

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/DecoupleSelfReloading.cs:
```diff
@@ -0,0 +1,86 @@
+			foreach (var pool in ammoPools)
+			{
+				var selfReloads = pool.LastChildMatching("SelfReloads");
+				if (selfReloads == null || !FieldLoader.GetValue<bool>("SelfReloads", selfReloads.Value.Value))
```
Likewise here with `NodeValue`.  You should never need to use `.Value.Value` directly in the update rules.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/DecoupleSelfReloading.cs:
```diff
@@ -0,0 +1,86 @@
+				{
+					var rd = selfReloadDelay;
+					rd.RenameKeyPreservingSuffix("Delay");
+					reloadOnCond.Value.Nodes.Add(rd);
```
We probably want to add an `AddNode` overload that accepts a MiniYamlNode for consistency, then avoid using `.Value.Nodes` in these cases.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/DecoupleSelfReloading.cs:
```diff
@@ -0,0 +1,86 @@
+				var reloadCount = pool.LastChildMatching("ReloadCount");
+				if (reloadCount != null)
+				{
+					var rc = reloadCount;
```
This appears to be redundant.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/DecoupleSelfReloading.cs:
```diff
@@ -0,0 +1,86 @@
+				var selfReloadDelay = pool.LastChildMatching("SelfReloadDelay");
+				if (selfReloadDelay != null)
+				{
+					var rd = selfReloadDelay;
```
This appears to be redundant.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/DecoupleSelfReloading.cs:
```diff
@@ -0,0 +1,86 @@
+				var rearmSound = pool.LastChildMatching("RearmSound");
+				if (rearmSound != null)
+				{
+					var rs = rearmSound;
```
This appears to be redundant.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/RemovePlayerPaletteTileset.cs:
```diff
@@ -0,0 +1,42 @@
+				ppfct.AddNode("Tileset", "");
+				ppfct.RenameKeyPreservingSuffix("PaletteFromFile");
+				yield return ppfct.Location + ": The trait 'PlayerPaletteFromCurrentTileset'\n" +
+					"has been removed. Use 'PaletteFromFile' with a Tileset filter.";
```
"has been replaced by 'PaletteFromFile'.  The trait has been renamed for you, but you will need to update the definition to specify the correct filename and tileset filters"

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/ScaleDefaultModHealth.cs:
```diff
@@ -0,0 +1,118 @@
+
+namespace OpenRA.Mods.Common.UpdateRules.Rules
+{
+	public class ScaleDefaultModHealth : UpdateRule
```
I'm not sure how useful this is as it is currently written - we do need to have rules keyed on mod ID for the resource center, but this won't run at all for other mods.  It would be better if we could have a base `ScaleModHealth` superclass that takes the scaling factor in the ctor, a pair of `ScaleModHealthBy10` and `ScaleModHealthBy100` subclasses that can be called manually by any mod, and then a `ScaleDefaultModHealth` subclass that then uses `ModScales` to pick the scaling value or skips if non-default.

abcdefg30 committed:
Add five more update rules

abcdefg30 committed:
Add ScaleDefaultHealth and subclass update rules

abcdefg30 commented:
Updated. Didn't test the rules yet, though.

abcdefg30 committed:
Add four more update rules

MustaphaTR commented on OpenRA.Mods.Common/UpdateRules/Rules/ScaleModHealth.cs:
```diff
@@ -0,0 +1,111 @@
+		{
+			{ "Health", new[] { Pair.New("HP", string.Empty) } },
+			{ "SelfHealing", new[] { Pair.New("Step", "500") } },
+			{ "RepairsUnits", new[] { Pair.New("HP", "1000") } },
```
This should be HpPerStep.

pchote commented:
Can you please take a look at #15090 and amend these to match?

pchote commented:
Ping @abcdefg30

reaperrr commented on OpenRA.Mods.Common/UpdateRules/UpdatePath.cs:
```diff
@@ -34,6 +34,19 @@ public class UpdatePath
+				new CapturableChanges(),
+				new DecoupleSelfReloading(),
+				new RemovePlayerPaletteTileset(),
+				new RemoveWithReloadingSpriteTurret(),
```
This must be a copy-paste error, this rule shouldn't be in this update path.

reaperrr committed:
Add ScaleSupportPowerSecondsToTicks

reaperrr commented on OpenRA.Mods.Common/UpdateRules/Rules/DecoupleSelfReloading.cs:
```diff
@@ -0,0 +1,83 @@
+			get
+			{
+				return "'SelfReloads', 'SelfReloadDelay', 'ReloadCount', 'ResetOnFire'\n" +
+					"and 'RearmSound' were renamed and moved to a new 'ReloadAmmoPool' trait.";
```
Neither Name above, nor this actually mentions they were moved from `AmmoPool`. Maybe just insert `from 'AmmoPool'` between the "move" and "to" here.

abcdefg30 replied:
Added.

reaperrr commented on OpenRA.Mods.Common/UpdateRules/Rules/DecoupleSelfReloading.cs:
```diff
@@ -0,0 +1,83 @@
+			foreach (var pool in ammoPools)
+			{
+				var selfReloads = pool.LastChildMatching("SelfReloads");
+				if (selfReloads == null || selfReloads.NodeValue<bool>())
```
If `.NodeValue<bool>()` simply returns the value, isn't this missing an `!` in front, as we only want to continue here if the the pool is **not** self-reloading?

abcdefg30 replied:
Ops, nice catch. Thanks.

reaperrr commented on OpenRA.Mods.Common/UpdateRules/Rules/DecoupleSelfReloading.cs:
```diff
@@ -0,0 +1,83 @@
+				{
+					selfReloadDelay.RenameKeyPreservingSuffix("Delay");
+					reloadOnCond.AddNode(selfReloadDelay);
+					pool.RemoveNodes("SelfReloads");
```
Why do you only remove "SelfReloads" if `selfReloadDelay != null`? It might have been using the internal default. I'd put this right after the 'continue' in the beginning of the loop, as we don't need it afterwards.

abcdefg30 replied:
I copied that from the original rule. Moved it out, though.

abcdefg30 commented:
Updated, all rules are ported now.

reaperrr commented:
I haven't really looked at them yet, but iirc the power refactor, IDisable removal etc. kind of depend on each other, so we have to make sure their rules don't screw up anything when run in arbitrary order.

pchote commented:
>  kind of depend on each other

It may make sense to merge these into one.

reaperrr commented:
By the way, I was actually working on an update rule for the Gate/Building split as well (I had planned that as separate PR which is why I didn't mention it earlier) and in the process developed two new UpdateUtils MiniYamlNode extensions locally:
`MoveNode(fromNode, toNode)` and `MoveAndRenameNode(fromNode, toNode, string newName)`.
Those can often save us 3 out of 4 or 4 out of 5 lines for each property that is moved to another trait (and possibly renamed in the process), because most of the time it allows us to drop curly brackets as well.
Particularly useful in cases like the Gate split, aim anim split and lobby checkbox rename rules.

Should I push them to this branch (along with updating the mentioned rules), or would you prefer that as a separate optimization PR?

pchote commented:
A separate optimization pr that tweaks any existing rules but doesn't introduce anything new would be best.

pchote commented:
What is the situation here?  I have lost an overview of the situation with these update rule PRs, and it seems that everything is mutually blocking everything else :cry: and that nothing can be reviewed.

reaperrr commented:
As far as I can tell, this PR "only" needs some of the separate inter-dependant update rules merged into single, larger ones.
- `MoveVisualBounds` needs to be updated to work on the assumption that `SplitSelectionAndRenderSize` hasn't been applied and the latter rule then needs to be removed (this should be relatively easy).
- `DropPauseAnimationWhenDisabled`, `ChangeCanPowerDown`, `ReplaceCanPowerDown`, `ReplaceRequiresPower` and `RemoveIDisable` seemingly need to be merged into a single monolithic `RemoveIDisableInterface` update rule, unfortunately. Probably best to check first in which order they were applied in the old system and then add them to the file in that order, each time checking for "what can be done to reduce redundancy/duplication".

The other rules look like they should work in isolation.

pchote commented:
We can keep the IDisable rules separate if that makes for cleaner code.  The rules should be applied in the order that they are specified in the update path.

abcdefg30 committed:
Add a 'ReplaceRequiresPower' update rule

abcdefg30 committed:
Add all remaining update rules

pchote commented on OpenRA.Mods.Common/UpdateRules/UpdateUtils.cs:
```diff
@@ -237,6 +237,12 @@ public static void Save(this YamlFileSet files)
 		}
 
+		/// <summary>Only sets the yaml key of the node. Does not preserve the @suffix!</summary>
+		public static void SetKey(this MiniYamlNode node, string newKey)
```
`RenameKeyPreservingSuffix` should be fine for most of the things you are using this for.  Do we really need another version?  If we really do, then this should be named `RenameKey` rather than `SetKey` for consistency.

abcdefg30 replied:
I used this for https://github.com/OpenRA/OpenRA/pull/15010/commits/b61bf340f1b6aeea21c31cfb5e68d6cd3984d2ab#diff-0bebab4e0f73ce1d3baeb516eb74e974R35 (`requiresPower.SetKey("GrantConditionOnPowerState@LOWPOWER");`). I'm not sure if using `RenameKeyPreservingSuffix` works in that case, as one could end up with a second suffix. I can think of another way if we don't want a second rename method. (Btw, I named it `SetKey` with the intention of making clear that the method doesn't preserve the suffix but overwrites it entirely.)

reaperrr replied:
I'm not sure what the advantage is over setting the desired key directly (`requiresPower.Key("GrantConditionOnPowerState@LOWPOWER";`)?
It doesn't even save anything in terms of line length (`Set` + `()` brackets = 5 bytes, ` = ` = 3 bytes).

pchote commented:
It looks like we are still missing rules for:
* Remove TargetWhenIdle/TargetWhenDamaged from AutoTarget [#13695]
* Refactor BurstDelay to BurstDelays [#13684]
* Types: support for InfiltrateFor* traits [#14311]

reaperrr commented:
> Remove TargetWhenIdle/TargetWhenDamaged from AutoTarget [#13695]

This one was already in 20171014 (I checked the source code).

reaperrr commented:
Added the last two missing rules (which need to be reviewed by someone else)

> Refactor BurstDelay to BurstDelays [#13684]
> Types: support for InfiltrateFor* traits [#14311]

and changed the 171014->180218 update path order to match the order the updates would be run in if you ran a full upgrade in the old system.

pchote commented:
Does this need to be updated to account for #15219?

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/WarnAboutInfiltrateForTypes.cs:
```diff
@@ -0,0 +1,66 @@
+
+		public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
+		{
+			var addActor = false;
```
This can simplify a lot, getting rid of `addActor`:

```csharp
foreach (var t in infiltrateForTraits)
{
	if (actorNode.LastChildMatching(t) != null)
	{
		infiltrateForLocations.Add(Tuple.Create(actorNode.Key, actorNode.Location.Filename));
		yield break;
	}
}
```

pchote commented:
The best course of action here may be to add a new update rule that sits at the start of the UpdatePath and simply logs a warning that the release-20171014 update path is "beta" quality and that issues can be expected and should be reported on GitHub or IRC.  We can then merge it, ship it, and fix it later.

reaperrr commented:
Sounds good to me.

> Does this need to be updated to account for #15219?

Yes, we should at least drop https://github.com/OpenRA/OpenRA/pull/15010/commits/534d2fb13b002982e18a098c12e3d6c9fd3d5be4 and adapt the rules that used it to #15219 instead.

reaperrr committed:
Add beta warning to legacy update path

reaperrr committed:
Move 20171014-20180218 update rules to subfolder

Also synced order of appearance with order they're applied in when running the full update path.

reaperrr commented:
Updated, also moved legacy rules into 20171014 subfolder as discussed on IRC.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/20171014/ReplaceRequiresPower.cs:
```diff
@@ -0,0 +1,43 @@
+			get
+			{
+				return "'RequiresPower' has been replaced with 'GrantConditionOnPowerState' which\n" +
+					"toggles a condition depending on the power state.\n Possible PowerStates are: Normal " +
```
Stray space between the newline and "Possible" looks odd:
```
  * ReplaceRequiresPower: Replace 'RequiresPower' with 'GrantConditionOnPowerState'
     'RequiresPower' has been replaced with 'GrantConditionOnPowerState' which
     toggles a condition depending on the power state.
      Possible PowerStates are: Normal (0 or positive), Low (negative but higher than 50% of required power)
     and Critical (below Low).
```

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/20171014/ReplaceRequiresPower.cs:
```diff
@@ -0,0 +1,43 @@
+			{
+				return "'RequiresPower' has been replaced with 'GrantConditionOnPowerState' which\n" +
+					"toggles a condition depending on the power state.\n Possible PowerStates are: Normal " +
+					"(0 or positive), Low (negative but higher than 50% of required power)\nand Critical (below Low).";
```
Would look better if the newline was moved from before the "and" to before the "50%"

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/20171014/LegacyBetaWarning.cs:
```diff
@@ -0,0 +1,40 @@
+
+		public override IEnumerable<string> AfterUpdate(ModData modData)
+		{
+			var message = "Due to time constraints and the legacy status of the included rules,\n" +
```
This message is currently repeated for all maps too!
Please add a flag to only display it the first time.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/20171014/RemoveOutOfAmmo.cs:
```diff
@@ -0,0 +1,68 @@
+				if (anyMatchingArmament)
+				{
+					pool.AddNode(ammoNoAmmo);
+					yield return "Aircraft returning to base is now triggered when all armaments are paused via condition.\n" +
```
This message is reoeated with identical messages for each match.  Please either report it only once for all actors, or add actor name / source file data.

pchote reviewed (commented):
Almost there IMO.

I noticed a few issues when printing the rule details and running them on the bleed version.

reaperrr commented:
Updated.

abcdefg30 commented:
Travis passes now.

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/20171014/DecoupleSelfReloading.cs:
```diff
@@ -0,0 +1,84 @@
+					pool.RemoveNodes("RearmSound");
+				}
+
+				actorNode.AddNode(reloadOnCond);
```
```
DecoupleSelfReloading: Replace 'SelfReloads' with 'ReloadAmmoPool'
   Updating mod... FAILED

   The automated changes for this rule were not applied because of an error.
   After the issue reported below is resolved you should run the updater
   with SOURCE set to DecoupleSelfReloading to retry these changes

   The exception reported was:
     System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
       at System.ThrowHelper.ThrowInvalidOperationException (System.ExceptionResource resource) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2017-12/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/throwhelper.cs:104 
       at System.Collections.Generic.List`1+Enumerator[T].MoveNextRare () [0x00013] in <bb7b695b8c6246b3ac1646577aea7650>:0 
       at System.Collections.Generic.List`1+Enumerator[T].MoveNext () [0x0004a] in <bb7b695b8c6246b3ac1646577aea7650>:0 
       at System.Linq.Enumerable+WhereListIterator`1[TSource].MoveNext () [0x0004e] in /Users/builder/jenkins/workspace/build-package-osx-mono/2017-12/external/bockbuild/builds/mono-x64/external/corefx/src/System.Linq/src/System/Linq/Where.cs:348 
       at OpenRA.Mods.Common.UpdateRules.Rules.DecoupleSelfReloading+<UpdateActorNode>c__Iterator0.MoveNext () [0x001ab] in /Users/paul/src/OpenRA/OpenRA.Mods.Common//UpdateRules/Rules/20171014/DecoupleSelfReloading.cs:32 
       at OpenRA.Mods.Common.UpdateRules.UpdateUtils+<ApplyTopLevelTransform>c__Iterator3.MoveNext () [0x00127] in /Users/paul/src/OpenRA/OpenRA.Mods.Common//UpdateRules/UpdateUtils.cs:220 
       at System.Collections.Generic.List`1[T].InsertRange (System.Int32 index, System.Collections.Generic.IEnumerable`1[T] collection) [0x000ea] in <bb7b695b8c6246b3ac1646577aea7650>:0 
       at System.Collections.Generic.List`1[T].AddRange (System.Collections.Generic.IEnumerable`1[T] collection) [0x00000] in <bb7b695b8c6246b3ac1646577aea7650>:0 
       at OpenRA.Mods.Common.UpdateRules.UpdateUtils.UpdateMod (OpenRA.ModData modData, OpenRA.Mods.Common.UpdateRules.UpdateRule rule, System.Collections.Generic.List`1[System.Tuple`3[OpenRA.FileSystem.IReadWritePackage,System.String,System.Collections.Generic.List`1[OpenRA.MiniYamlNode]]]& files, System.Collections.Generic.HashSet`1[T] externalFilenames) [0x00265] in /Users/paul/src/OpenRA/OpenRA.Mods.Common//UpdateRules/UpdateUtils.cs:173 
       at OpenRA.Mods.Common.UtilityCommands.UpdateModCommand.ApplyRules (OpenRA.ModData modData, System.Collections.Generic.IEnumerable`1[T] rules, System.Boolean skipMaps) [0x00081] in /Users/paul/src/OpenRA/OpenRA.Mods.Common//UtilityCommands/UpdateModCommand.cs:152 
```

This either needs to move outside the loop, or `ammoPools` needs to be `ToList()`ed so that you aren't enumerating the same collection you're modifying.

reaperrr replied:
Fixed by going the `ToList()` route.

pchote reviewed (commented):
One last crash, then I think this will be okay enough to ship:

pchote commented:
@abcdefg30: Will you want to give this a last look over before we merge, or shall we merge immediately once the issue above is fixed?

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/20171014/AddNukeLaunchAnimation.cs:
```diff
@@ -0,0 +1,46 @@
+
+				var sequence = activation.NodeValue<string>();
+				nuke.RemoveNode(activation);
+				actorNode.AddNode("WithNukeLaunchAnimation", "");
```
Same issue as before.  Discovered when running the update rules on D2K:
```
AddNukeLaunchAnimation: Add 'WithNukeLaunchAnimation' and remove 'NukePower.ActivationSequence'
   Updating mod... FAILED

   The automated changes for this rule were not applied because of an error.
   After the issue reported below is resolved you should run the updater
   with SOURCE set to AddNukeLaunchAnimation to retry these changes

   The exception reported was:
     System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
       at System.ThrowHelper.ThrowInvalidOperationException (System.ExceptionResource resource) [0x0000b] in <bb7b695b8c6246b3ac1646577aea7650>:0 
       at System.Collections.Generic.List`1+Enumerator[T].MoveNextRare () [0x00013] in <bb7b695b8c6246b3ac1646577aea7650>:0 
       at System.Collections.Generic.List`1+Enumerator[T].MoveNext () [0x0004a] in <bb7b695b8c6246b3ac1646577aea7650>:0 
       at System.Linq.Enumerable+WhereListIterator`1[TSource].MoveNext () [0x0004e] in <b29fa38c592649ee8d566c045a7d9307>:0 
       at OpenRA.Mods.Common.UpdateRules.Rules.AddNukeLaunchAnimation+<UpdateActorNode>c__Iterator0.MoveNext () [0x000a1] in <300dcefade434fd18783a65f41b5ec39>:0 
       at OpenRA.Mods.Common.UpdateRules.UpdateUtils+<ApplyTopLevelTransform>c__Iterator3.MoveNext () [0x00121] in <300dcefade434fd18783a65f41b5ec39>:0 
       at System.Collections.Generic.List`1[T].InsertRange (System.Int32 index, System.Collections.Generic.IEnumerable`1[T] collection) [0x000ea] in <bb7b695b8c6246b3ac1646577aea7650>:0 
       at System.Collections.Generic.List`1[T].AddRange (System.Collections.Generic.IEnumerable`1[T] collection) [0x00000] in <bb7b695b8c6246b3ac1646577aea7650>:0 
       at OpenRA.Mods.Common.UpdateRules.UpdateUtils.UpdateMod (OpenRA.ModData modData, OpenRA.Mods.Common.UpdateRules.UpdateRule rule, System.Collections.Generic.List`1[System.Tuple`3[OpenRA.FileSystem.IReadWritePackage,System.String,System.Collections.Generic.List`1[OpenRA.MiniYamlNode]]]& files, System.Collections.Generic.HashSet`1[T] externalFilenames) [0x00265] in <300dcefade434fd18783a65f41b5ec39>:0 
       at OpenRA.Mods.Common.UtilityCommands.UpdateModCommand.ApplyRules (OpenRA.ModData modData, System.Collections.Generic.IEnumerable`1[T] rules, System.Boolean skipMaps) [0x0007a] in <300dcefade434fd18783a65f41b5ec39>:0 
```

pchote replied:
I've checked for all other uses of `AddNode` in this PR and it looks like this is the last case to fix.

reaperrr replied:
Fixed.

pchote commented:
Taking this without a second :+1: so that we can move ahead with a playtest.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/56094626c9b334a1c553567b736b686665594c7b
